extensions/VisualEditor/wikia/modules/ve/ui/dialogs/ve.ui.WikiaInfoboxInsertDialog.js

```diff
@@ -155,26 +139,112 @@ ve.ui.WikiaInfoboxInsertDialog.prototype.insertInfoboxTemplate = function () {
 };
 
 /**
- * Add the infobox template names to the dialog's SelectWidget
- *
+ * @desc creates infobox item option widget
+ * @param {Object} data
+ * @returns {OO.ui.DecoratedOptionWidget}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.createInfoboxItemOptionWidget = function ( data ) {
+	return new OO.ui.DecoratedOptionWidget( {
+		data: data.title,
+		label:  data.title
+	} );
+};
+
+/**
+ * @desc creates select widget with infobox option widgets
+ * @param {OO.ui.DecoratedOptionWidget[]} items
+ * @returns {OO.ui.OoUiSelectWidget}
+ */
+ve.ui.WikiaInfoboxInsertDialog.prototype.createInfoboxSelectWidget = function ( items ) {
+	var select = new OO.ui.SelectWidget();
+
+	select.addItems( items );
+
+	return select;
+};
+
+/**
+ * @desc creates dialog content
  * @param {Object} data Response data from API
+ * @returns {Promise}
  */
-ve.ui.WikiaInfoboxInsertDialog.prototype.showResults = function ( data ) {
-	var i,
-		items = [],
-		infoboxes = data.query ? data.query.allinfoboxes : [];
-
-	if ( infoboxes.length > 0 ) {
-		for ( i = 0; i < infoboxes.length; i++ ) {
-			items.push(
-				new OO.ui.DecoratedOptionWidget({
-					data: infoboxes[i].title,
-					label:  infoboxes[i].title
-				})
-			);
-		}
-		this.select.addItems( items );
+ve.ui.WikiaInfoboxInsertDialog.prototype.createDialogContent = function ( data ) {
+	var deferred = $.Deferred(),
+		infoboxTemplatesData = data.query ? data.query.allinfoboxes : [],
+		infoboxTemplateOptionWidgets = infoboxTemplatesData.map( this.createInfoboxItemOptionWidget );
+
+	if ( infoboxTemplateOptionWidgets.length > 0 ) {
+		this.select = this.createInfoboxSelectWidget( infoboxTemplateOptionWidgets );
+
+		this.select.connect( this, {
+			select: 'onInfoboxTemplateSelect'
+		} );
+
+		deferred.resolve( this.select.$element );
+	} else {
+		// creates empty state content
+		this.getUnconvertedInfoboxes().then(
+			( function ( data ) {
+				deferred.resolve( this.createEmptyStateContent( data ).bind( this ) );
+			} ).bind( this )
+		);
```

Comment:
I find those few lines above little bit hard to follow, if you want to do one thing (getUnconvertedInfoboxes) then do other thing (createEmptyStateContent) and then do another thing (resolve) why not just use then two times? Remember that the output data from first then will be carried over to second then. I actually created this simple example for you so you can see how it works exactly: https://jsfiddle.net/Inez/ggu74j6c/